### ISSUE TITLE
awscli fails when running the as part of a Daemon which does not have a stdin

### ISSUE THREAD

**Issue by markdessain** [aws/aws-cli#2518]
Currently I am running jobs via the Apache Airflow Scheduler (https://airflow.incubator.apache.org/)

It runs background jobs as a Daemon with only stdout and stderr attached, but no stdin - https://github.com/apache/incubator-airflow/blob/master/airflow/bin/cli.py#L824-L829 / https://pagure.io/python-daemon/blob/master/f/daemon/daemon.py#_255

I have discovered that while running under Python3 the awscli will only work if the stdin is set due to - https://github.com/aws/aws-cli/blob/b5fe9872abc6006f2a3093d4d079d23024bc3e0a/awscli/compat.py#L75

Currently it raises `AttributeError: 'NoneType' object has no attribute 'buffer'` as stdin is `None`

-------------

A full example showing it working under Python2 but not Python3 can be found here:

```

# Works
# ========
# mkvirtualenv awscli_python2 -p python
# pip install awscli==1.11.68
# pip install python-daemon==2.1.1
# python aws_stdin.py

# Does not work
# ========
# mkvirtualenv awscli_python3 -p python3
# pip install awscli==1.11.68
# pip install python-daemon==2.1.1
# python aws_stdin.py


import os
import sys
import daemon
from subprocess import Popen, STDOUT, PIPE

stdout = open('out', 'w+')

ctx = daemon.DaemonContext(
    stdout=stdout
)

with ctx:
    sp = Popen(['aws'], stdout=PIPE, stderr=STDOUT)

    for line in iter(sp.stdout.readline, b''):
        print(line)

```

---------

```
b'Traceback (most recent call last):\n'
b'  File "/Users/markdessain/Envs/aws_stdin/bin/aws", line 19, in <module>\n'
b'    import awscli.clidriver\n'
b'  File "/Users/markdessain/Envs/aws_stdin/lib/python3.5/site-packages/awscli/clidriver.py", line 26, in <module>\n'
b'    from awscli.formatter import get_formatter\n'
b'  File "/Users/markdessain/Envs/aws_stdin/lib/python3.5/site-packages/awscli/formatter.py", line 19, in <module>\n'
b'    from awscli.table import MultiTable, Styler, ColorizedStyler\n'
b'  File "/Users/markdessain/Envs/aws_stdin/lib/python3.5/site-packages/awscli/table.py", line 19, in <module>\n'
b'    from awscli.compat import six\n'
b'  File "/Users/markdessain/Envs/aws_stdin/lib/python3.5/site-packages/awscli/compat.py", line 75, in <module>\n'
b'    binary_stdin = sys.stdin.buffer\n'
b"AttributeError: 'NoneType' object has no attribute 'buffer'\n"
```

**Comment by kyleknap** [comment]
Thanks for the steps to reproduce this and the PR. That was very helpful. I was able to reproduce it as well.

Digging into this a bit more, it looks like it is because of the use of ``python-daemon`` package. When I first tried, I actually used just the normal ``Popen`` and no daemon context which worked just fine. Looking into the ``python-daemon`` package it looks like if you do not set ``stdin`` in the ``DaemonContext`` it sets ``sys.stdin`` to ``None`` in the process being ran, which is why you get that error. To confirm this, I updated your code:
```python
import os
import sys
import daemon
from subprocess import Popen, STDOUT, PIPE

stdout = open('out', 'w+')

ctx = daemon.DaemonContext(
    stdin=sys.stdin,
    stdout=stdout
)

with ctx:
    sp = Popen(['aws'], stdout=PIPE, stderr=STDOUT)

    for line in iter(sp.stdout.readline, b''):
        print(line)
```
And that then works as expected. Would it be possible to set a ``stdin`` somewhere in your script? I am little hesitant to pull in the patch for this because it is mainly to support this use case with a third party package. If this was happening just straight up with Python's standard library I would be more inclined to accommodate it. Furthermore, there are other places in the CLI code base where ``sys.stdin`` is used so if you end up hitting that code path, you may start seeing similar errors even if the patch gets pulled in to handle it at that location.

**Comment by markdessain** [comment]
I was not aware that the `sys.stdin` was used else where, so your answer makes sense.

Unfortunately its not part of some code that I've written but in the Airflow Library. I can get around it by using boto3 (which I probably should have done in the first place).

So I'll leave it as not a pressing concern but glad you recognise the difference between Python2 and Python3 incase anyone else falls into the same problem. Thanks again for looking into this.

**Comment by mobrien** [comment]
Just hit this as well running awscli in a detached process with no stdin.

**Comment by emurphy** [comment]
@stealthycoin and @kyleknap can you please reopen this and pull in the patch from #2519?

The python docs are clear that a library should not rely on `sys.stdin.buffer`:

from https://docs.python.org/3/library/sys.html:
Note To write or read binary data from/to the standard streams, use the underlying binary buffer object. For example, to write bytes to stdout, use sys.stdout.buffer.write(b'abc').
However, if you are writing a library (and do not control in which context its code will be executed), be aware that the standard streams may be replaced with file-like objects like io.StringIO which do not support the buffer attribute.

Have also manually patched the proposed change from #2519, and it works for our use case.
